# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/mediacontrol/MediaStatusManager.cpp
# Commit: 37510e9b92b2
# Full Hash: 37510e9b92b2a7479b45fffec345286a67c3c891
# Author: alwu <alwu@mozilla.com>
# Date: 2020-06-17 03:41:56
# Description:
#   Bug 1644479 - replace assertions with early returns. r=chunmin
#   
#   Still can not figure out how did we enable the same action twice, because we would only update the action change when the action handler is set from `null` to `something` or from `something` to `nothing` [1]. In addition, the browsing context Id for a media session is always unchanged, so it's guarantee to set the action on the correct `MediaSessionInfo`.
#   
#   However, in order to avoid having more crashes, I have to change the assertions to early returns and add logs to help us be aware of this issue if it happens again.
# ==============================================================================

diff -r 632c0f1ee377 -r 37510e9b92b2 dom/media/mediacontrol/MediaStatusManager.cpp
--- a/dom/media/mediacontrol/MediaStatusManager.cpp	Tue Jun 16 18:58:54 2020 +0000
+++ b/dom/media/mediacontrol/MediaStatusManager.cpp	Tue Jun 16 19:05:02 2020 +0000
@@ -310,8 +310,11 @@
     return;
   }
   MediaSessionInfo* info = mMediaSessionInfoMap.GetValue(aBrowsingContextId);
-  MOZ_DIAGNOSTIC_ASSERT(!info->IsActionSupported(aAction),
-                        "Action has already been enabled!");
+  if (info->IsActionSupported(aAction)) {
+    LOG("Action '%s' has already been enabled for context %" PRIu64,
+        ToMediaSessionActionStr(aAction), aBrowsingContextId);
+    return;
+  }
   LOG("Enable action %s for context %" PRIu64, ToMediaSessionActionStr(aAction),
       aBrowsingContextId);
   info->EnableAction(aAction);
@@ -324,8 +327,11 @@
     return;
   }
   MediaSessionInfo* info = mMediaSessionInfoMap.GetValue(aBrowsingContextId);
-  MOZ_DIAGNOSTIC_ASSERT(info->IsActionSupported(aAction),
-                        "Action hasn't been enabled yet!");
+  if (!info->IsActionSupported(aAction)) {
+    LOG("Action '%s' hasn't been enabled yet for context %" PRIu64,
+        ToMediaSessionActionStr(aAction), aBrowsingContextId);
+    return;
+  }
   LOG("Disable action %s for context %" PRIu64,
       ToMediaSessionActionStr(aAction), aBrowsingContextId);
   info->DisableAction(aAction);
